redis/src/sentinel.rs

```diff
@@ -0,0 +1,795 @@
+        }
+    }
+    false
+}
+
+/// Async version of [find_valid_master].
+#[cfg(feature = "aio")]
+async fn async_find_valid_master(
+    masters: Vec<HashMap<String, String>>,
+    service_name: &str,
+    node_connection_info: &SentinelNodeConnectionInfo,
+) -> RedisResult<ConnectionInfo> {
+    for (ip, port) in valid_addrs(masters, |m| is_master_valid(m, service_name)) {
+        let connection_info = node_connection_info.create_connection_info(ip, port);
+        if async_check_role(&connection_info, "master").await {
+            return Ok(connection_info);
+        }
+    }
+
+    fail!((
+        ErrorKind::MasterNameNotFound,
+        "Master with given name not found in sentinel",
+    ))
+}
+
+fn get_valid_replicas_addresses(
+    replicas: Vec<HashMap<String, String>>,
+    node_connection_info: &SentinelNodeConnectionInfo,
+) -> Vec<ConnectionInfo> {
+    valid_addrs(replicas, is_replica_valid)
+        .map(|(ip, port)| node_connection_info.create_connection_info(ip, port))
+        .filter(|connection_info| check_role(connection_info, "slave"))
+        .collect()
+}
+
+#[cfg(feature = "aio")]
+async fn async_get_valid_replicas_addresses<'a>(
+    replicas: Vec<HashMap<String, String>>,
+    node_connection_info: &SentinelNodeConnectionInfo,
+) -> Vec<ConnectionInfo> {
+    async fn is_replica_role_valid(connection_info: ConnectionInfo) -> Option<ConnectionInfo> {
+        if async_check_role(&connection_info, "slave").await {
+            Some(connection_info)
+        } else {
+            None
+        }
+    }
+
+    futures_util::stream::iter(valid_addrs(replicas, is_replica_valid))
+        .map(|(ip, port)| node_connection_info.create_connection_info(ip, port))
+        .filter_map(is_replica_role_valid)
+        .collect()
+        .await
+}
+
+#[cfg(feature = "aio")]
+async fn async_reconnect(
+    connection: &mut Option<AsyncConnection>,
+    connection_info: &ConnectionInfo,
+) -> RedisResult<()> {
+    let sentinel_client = Client::open(connection_info.clone())?;
+    let new_connection = sentinel_client.get_async_connection().await?;
+    connection.replace(new_connection);
+    Ok(())
+}
+
+#[cfg(feature = "aio")]
+async fn async_try_single_sentinel<T: FromRedisValue>(
+    cmd: Cmd,
+    connection_info: &ConnectionInfo,
+    cached_connection: &mut Option<AsyncConnection>,
+) -> RedisResult<T> {
+    if cached_connection.is_none() {
+        async_reconnect(cached_connection, connection_info).await?;
+    }
+
+    let result = cmd.query_async(cached_connection.as_mut().unwrap()).await;
+
+    if let Err(err) = result {
+        if err.is_connection_dropped() || err.is_io_error() {
+            async_reconnect(cached_connection, connection_info).await?;
+            cmd.query_async(cached_connection.as_mut().unwrap()).await
+        } else {
+            Err(err)
+        }
+    } else {
+        result
+    }
+}
+
+// async methods
+#[cfg(feature = "aio")]
```

Comment:
minor:
Maybe move this to under the sync implementation, since some of the comments reference functions from there?

Also, please be consistent, and make sure that any function comment that can reference the sync implementation does so. you don't have to add a comment just for the reference, but if you already have one, make sure that the repetition of content is minimal.

Reply:
I moved it below, makes sense. But I've also removed the references, and just expanded the comment above, is that ok? I've also kept the doc comments in the public methods, but they're exactly the same as the sync versions. You think in this case we should just reference the sync versions to avoid repetition? I'm not sure what would be the best practice in this case.